### PR TITLE
[CI/Build] Support compilation with local cutlass path (#10423)

### PR DESCRIPTION
**Pull request by wchen61**

### CODE REVIEW

**Review by tlrmchlsmth — COMMENTED** [review]
looks good! I'll try it out

Great timing with this PR, since we are about to update CUTLASS to a pre-release version which will unfortunately mean we have to turn off GIT_SHALLOW

**Review by mgoin — COMMENTED** [review]
Could you mention this in the documentation somewhere? Maybe in the build from source section https://docs.vllm.ai/en/latest/getting_started/installation.html#build-from-source

**Comment by wchen61** [comment]
> Could you mention this in the documentation somewhere? Maybe in the build from source section https://docs.vllm.ai/en/latest/getting_started/installation.html#build-from-source

hi，I updated the doc, but I don't know how to preview the changes. Please review it.

**Comment by jeejeelee** [comment]
image

File: docs/source/getting_started/installation.rst [hunk 1]
```diff
@@ -170,6 +170,18 @@ To build vLLM using an existing PyTorch installation:
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Currently, before starting the build process, vLLM fetches cutlass code from GitHub. However, there may be scenarios where you want to use a local version of cutlass instead.
+To achieve this, you can set the environment variable CUTLASS_SRC_DIR to point to your local cutlass directory.
+
+.. code-block:: console
+
+    $ git clone https://github.com/vllm-project/vllm.git
+    $ cd vllm
+    $ CUTLASS_SRC_DIR=/path/to/cutlass pip install -e .
+
```
Comment (tlrmchlsmth) [inline docs/source/getting_started/installation.rst:183]: I found that relative paths don't work for specifying the CUTLASS_SRC_DIR. 

This is a pretty easy thing for users to get tripped up on, so could you document that absolute paths must be specified (or get relative paths working if you see a way to do that)?

File: CMakeLists.txt [hunk 2]
```diff
@@ -206,7 +206,15 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
   # Set CUTLASS_REVISION manually -- its revision detection doesn't work in this case.
   set(CUTLASS_REVISION "v3.5.1" CACHE STRING "CUTLASS revision to use")
 
-  FetchContent_Declare(
+  # Use the specified CUTLASS source directory for compilation if CUTLASS_SRC_DIR is provided
+  if (DEFINED ENV{CUTLASS_SRC_DIR})
+    set(CUTLASS_SRC_DIR $ENV{CUTLASS_SRC_DIR})
+  endif()
+
+  if(CUTLASS_SRC_DIR)
```
Comment (tlrmchlsmth) [inline CMakeLists.txt:214]: I think we should probably name the env variable `VLLM_CUTLASS_SRC_DIR` for namespacing

File: docs/source/getting_started/installation.rst (hunk 1 above)
Comment (wchen61) [reply docs/source/getting_started/installation.rst]: Great idea, I've implemented a logic to convert relative paths to absolute paths directly in the CMake script. Please review it to see if it's feasible.

**Review by mgoin — APPROVED** [review]
LGTM thanks!

**Review by tlrmchlsmth — APPROVED** [review]
Nice, thanks for getting local paths to work! Tried it out and works for me

**Comment by wchen61** [comment]
> Please merge from main to fix the CI failures.

Hi, I've updated to the latest version, but there are still some failures. I'm not sure if they are related to my patch. Could you take a look?

**Pull request merged**